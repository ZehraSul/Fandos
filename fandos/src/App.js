import "./App.css";
import { Routes, Route, Navigate, BrowserRouter } from "react-router-dom";
import Login from "./Components/Login";
import Menu from "./Components/Menu";
import Register from "./Components/Register";
import LandingPage from "./Components/LandingPage";
import { useState, useEffect } from "react";
import BasicNavbar from "./Components/BasicNavbar";
import { FANDOS_API_URL } from "./config/config";
import OrderHistory from "./Components/OrderHistory";
import Cart from "./Components/Cart";
import AdminNavbar from "./Components/AdminNavbar";
import Dashboard from "./Components/Dashboard";

function App() {
  let [isLoggedIn, setIsLoggedIn] = useState(false);
  let [userToken, setUserToken] = useState("");
  let [prevOrders, setPrevOrders] = useState([]);
  let [displayMenu, setDisplayMenu] = useState([]);
  let [cartItems, setCartItems] = useState([]);
  let [userType, setUserType] = useState("");

  /* Using useEffect to make the initial call to my api to display menu*/
  useEffect(() => {
    // if user is already logged in, get the token and set it as userToken
    let token = localStorage.getItem("token");
    let type = localStorage.getItem("userType");
    if (token) {
      setUserToken(token);
      setIsLoggedIn(true);
      setUserType(type);
    }
    // if user is logged in display all their tasks
    if (userToken) {
      fetch(`${FANDOS_API_URL}/menu/displayAll`, {
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          // Uses token for authorization
          Authorization: "Bearer " + userToken,
        },
      })
        .then((res) => {
          return res.json();
        })
        .then(
          (data) => {
            setDisplayMenu(data);
          },
          (err) => {
            console.log(err);
          }
        );
    }
  }, [userToken]);

  return (
    <div className="App">
      <BrowserRouter>
        {userType === "admin" ? (
          <AdminNavbar
            setUserToken={setUserToken}
            isLoggedIn={isLoggedIn}
            setIsLoggedIn={setIsLoggedIn}
            setUserType={setUserType}
            setCartItems={setCartItems}
            setPrevOrders={setPrevOrders}
          />
        ) : (
          <BasicNavbar
            setUserToken={setUserToken}
            isLoggedIn={isLoggedIn}
            setIsLoggedIn={setIsLoggedIn}
            setCartItems={setCartItems}
            setPrevOrders={setPrevOrders}
            setUserType={setUserType}
          />
        )}

        <Routes>
          <Route
            path="/"
            element={
              <LandingPage
                setUserToken={setUserToken}
                isLoggedIn={isLoggedIn}
                setIsLoggedIn={setIsLoggedIn}
              />
            }
          />
          <Route
            path="menu"
            element={
              <PrivateRoute>
                <Menu
                  displayMenu={displayMenu}
                  cartItems={cartItems}
                  setCartItems={setCartItems}
                />
              </PrivateRoute>
            }
          />
          <Route
            path="login"
            element={
              <Login
                setIsLoggedIn={setIsLoggedIn}
                setUserToken={setUserToken}
                setUserType={setUserType}
              />
            }
          />
          <Route
            path="register"
            element={
              <Register
                setIsLoggedIn={setIsLoggedIn}
                setUserToken={setUserToken}
                setCartItems={setCartItems}
              />
            }
          />
          <Route
            path="orderhistory"
            element={
              <PrivateRoute>
                <OrderHistory
                  prevOrders={prevOrders}
                  setPrevOrders={setPrevOrders}
                  setUserToken={setUserToken}
                  userToken={userToken}
                />
              </PrivateRoute>
            }
          />
          <Route
            path="cart"
            element={
              <PrivateRoute>
                <Cart
                  userToken={userToken}
                  setUserToken={setUserToken}
                  cartItems={cartItems}
                  setCartItems={setCartItems}
                  prevOrders={prevOrders}
                  setPrevOrders={setPrevOrders}
                />
              </PrivateRoute>
            }
          />
          <Route
            path="dashboard"
            element={
              <AdminRoute>
                <Dashboard
                  displayMenu={displayMenu}
                  setDisplayMenu={setDisplayMenu}
                  setUserToken={setUserToken}
                  isLoggedIn={isLoggedIn}
                  setIsLoggedIn={setIsLoggedIn}
                />
              </AdminRoute>
            }
          />
        </Routes>
      </BrowserRouter>
    </div>
  );
}

function useAuth() {
  let token = localStorage.getItem("token");
  return token;
}

function isAdmin() {
  let userType = localStorage.getItem("userType");
  return userType && userType === "admin";
}

function AdminRoute({ children }) {
  return useAuth() && isAdmin() ? children : <Navigate to="/" />;
}

function PrivateRoute({ children }) {
  return useAuth() ? children : <Navigate to="/login" />;
}

export default App;
