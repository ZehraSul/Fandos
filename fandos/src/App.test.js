import renderer from "react-test-renderer";
import App from "./App";

// Snapshot test
test("App renders correctly", () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
