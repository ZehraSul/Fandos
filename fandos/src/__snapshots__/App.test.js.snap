// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="App"
>
  <div>
    <div
      className="navbar-custom nav"
      onKeyDown={[Function]}
    >
      <div
        className="nav-item"
      >
        <a
          className="nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          href="/"
          onClick={[Function]}
        >
          Fandos
        </a>
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "margin": "auto",
          "paddingTop": "50px",
          "width": "50%",
        }
      }
    >
      <div
        className="carousel slide carousel-light"
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="carousel-indicators"
        >
          <button
            aria-current={true}
            aria-label="Slide 1"
            className="active"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
          <button
            aria-current={false}
            aria-label="Slide 2"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
          <button
            aria-current={false}
            aria-label="Slide 3"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
        </div>
        <div
          className="carousel-inner"
        >
          <div
            className="active carousel-item"
          >
            <img
              alt="First slide"
              className="d-block w-100"
              height="450px"
              src="/images/full-chicken.jpg"
            />
          </div>
          <div
            className="carousel-item"
          >
            <img
              alt="Second slide"
              className="d-block w-100"
              height="450px"
              src="/images/half-chicken.jpg"
            />
          </div>
          <div
            className="carousel-item"
          >
            <img
              alt="Third slide"
              className="d-block w-100"
              height="450px"
              src="/images/quarter-chicken.jpg"
            />
          </div>
        </div>
        <a
          className="carousel-control-prev"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            aria-hidden="true"
            className="carousel-control-prev-icon"
          />
          <span
            className="visually-hidden"
          >
            Previous
          </span>
        </a>
        <a
          className="carousel-control-next"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            aria-hidden="true"
            className="carousel-control-next-icon"
          />
          <span
            className="visually-hidden"
          >
            Next
          </span>
        </a>
      </div>
    </div>
    <div
      className="py-5"
      style={
        Object {
          "margin": "auto",
          "width": "50%",
        }
      }
    >
      <h2>
        Want some chicken? 
      </h2>
      <a
        className="btn btn-danger"
        disabled={false}
        href="/login"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Login
      </a>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <div>
    <div
      className="navbar-custom nav"
      onKeyDown={[Function]}
    >
      <div
        className="nav-item"
      >
        <a
          className="nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          href="/"
          onClick={[Function]}
        >
          Fandos
        </a>
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "margin": "auto",
          "paddingTop": "50px",
          "width": "50%",
        }
      }
    >
      <div
        className="carousel slide carousel-light"
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="carousel-indicators"
        >
          <button
            aria-current={true}
            aria-label="Slide 1"
            className="active"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
          <button
            aria-current={false}
            aria-label="Slide 2"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
          <button
            aria-current={false}
            aria-label="Slide 3"
            data-bs-target=""
            onClick={[Function]}
            type="button"
          />
        </div>
        <div
          className="carousel-inner"
        >
          <div
            className="active carousel-item"
          >
            <img
              alt="First slide"
              className="d-block w-100"
              height="450px"
              src="/images/full-chicken.jpg"
            />
          </div>
          <div
            className="carousel-item"
          >
            <img
              alt="Second slide"
              className="d-block w-100"
              height="450px"
              src="/images/half-chicken.jpg"
            />
          </div>
          <div
            className="carousel-item"
          >
            <img
              alt="Third slide"
              className="d-block w-100"
              height="450px"
              src="/images/quarter-chicken.jpg"
            />
          </div>
        </div>
        <a
          className="carousel-control-prev"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            aria-hidden="true"
            className="carousel-control-prev-icon"
          />
          <span
            className="visually-hidden"
          >
            Previous
          </span>
        </a>
        <a
          className="carousel-control-next"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            aria-hidden="true"
            className="carousel-control-next-icon"
          />
          <span
            className="visually-hidden"
          >
            Next
          </span>
        </a>
      </div>
    </div>
    <div
      className="py-5"
      style={
        Object {
          "margin": "auto",
          "width": "50%",
        }
      }
    >
      <h2>
        Want some chicken? 
      </h2>
      <a
        className="btn btn-danger"
        disabled={false}
        href="/login"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Login
      </a>
    </div>
  </div>
</div>
`;
